synapse/push/httppusher.py

```diff
@@ -358,80 +379,115 @@ async def _build_notification_dict(
+            "sender": event.user_id,
+            "prio": priority,
+            "counts": {
+                "unread": badge,
+                # 'missed_calls': 2
+            },
         }
         if event.type == "m.room.member" and event.is_state():
-            d["notification"]["membership"] = event.content["membership"]
-            d["notification"]["user_is_target"] = event.state_key == self.user_id
+            content["membership"] = event.content["membership"]
+            content["user_is_target"] = event.state_key == self.user_id
         if self.hs.config.push.push_include_content and event.content:
-            d["notification"]["content"] = event.content
+            content["content"] = event.content
 
         # We no longer send aliases separately, instead, we send the human
         # readable name of the room, which may be an alias.
         if "sender_display_name" in ctx and len(ctx["sender_display_name"]) > 0:
-            d["notification"]["sender_display_name"] = ctx["sender_display_name"]
+            content["sender_display_name"] = ctx["sender_display_name"]
         if "name" in ctx and len(ctx["name"]) > 0:
-            d["notification"]["room_name"] = ctx["name"]
+            content["room_name"] = ctx["name"]
+
+        return (content, tweaks)
 
-        return d
+    def _build_notification_dict(
+        self,
+        content: JsonDict,
+        tweaks: Optional[JsonMapping],
+        default_payload: Optional[JsonMapping] = None,
+    ) -> JsonDict:
+        """Build a full notification from the content of it, as specified by the
+        Push Gateway spec. It wraps the content in a top level `notification`
+        element and put relevant device info in the `devices` section.
+
+        Args:
+            content: the content
+            tweaks: tweaks to add into the `devices` section
+            default_payload: default payload to add in `devices[0].data.default_payload`.
+            This will be merged (and override if some matching values already exist there)
+            with existing `default_payload`.
+
+        Returns:
+            a full notification that can be send to the registered push gateway.
+        """
+        content = content.copy()
+
+        data = self.data_minus_url.copy()
+        if default_payload:
+            if "default_payload" not in data:
+                data["default_payload"] = {}
+            data["default_payload"].update(default_payload)
+
+        device = {
+            "app_id": self.app_id,
+            "pushkey": self.pushkey,
+            "pushkey_ts": int(self.pushkey_ts / 1000),
+            "data": data,
+        }
+        if tweaks:
+            device["tweaks"] = tweaks
+
+        content["devices"] = [device]
+
+        return {"notification": content}
 
     async def dispatch_push(
-        self, event: EventBase, tweaks: Dict[str, bool], badge: int
+        self,
+        content: JsonDict,
+        tweaks: Optional[JsonMapping] = None,
```

Comment:
This significantly expands the type of `tweaks` -- is that on purpose? It seems incorrect.

Reply:
I think the original type hint is wrong. The values can at least have `bool` and `str` from what I understand from the code and the example in the spec. The spec itself doesn't define `tweaks` :/ .
`JsonMapping` is probably too broad, I just checked and I should probably use `Mapping[str, SimpleJsonValue]` I think.

Reply:
I'm pretty sure we don't do validation of what tweaks you set so this could be anything, but I'm not 100% sure. Where do we build this dictionary that makes it safe to make this change?

Reply:
So original code was using `Dict[str, bool]`, and I am broadening the type, because clearly the type hint is wrong: [tweak_for_actions](https://github.com/matrix-org/synapse/blob/develop/synapse/push/httppusher.py#L59) is returning a `Dict[str, Any]`.
I am not really understanding your point sorry :) I was using `JsonMapping` but moved to `Mapping[str, SimpleJsonValue]` after your concern about being too broad, but I think we should use `JsonMapping` and I am going to change that again since it seems you agree :) ?

Reply:
Alright, I see what you're saying. I agree that `JsonMapping` is likely the correct thing to use. To be clear -- I didn't know where this value came from, so pointing to `tweak_for_actions` is helpful in understanding what generates this dictionary.